source/common/upstream/upstream_impl.h

```diff
@@ -156,8 +153,7 @@ class HostDescriptionImpl : virtual public HostDescription {
   std::shared_ptr<envoy::api::v2::core::Metadata> metadata_ ABSL_GUARDED_BY(metadata_mutex_);
   const envoy::api::v2::core::Locality locality_;
   Stats::StatNameManagedStorage locality_zone_stat_name_;
-  Stats::IsolatedStoreImpl stats_store_;
-  HostStats stats_;
+  mutable HostStats stats_;
```

Comment:
qq: what was the reasoning behind the mutable and const to non-const switch here? Presumably because we were previously violating the constness anyway and this makes it more explicit? It would be pretty cool to not do this. I assume you looked and it's not easy to fix this right now because we modify stats already though a const host?

Reply:
Before this change HostStats contained non-const references to counter objects, while now it holds the counters themselves as members.

Consider host.stats().rq_success_.inc() before and after that change:
before: The HostStats returned by Host::stats is const, but rq_success_ is not const so the call to inc() is fine.
after: if HostStats is const its members are also const, attempts to call inc() on the const counter object will fail compilation.

We could instead make the counters and gauges inside HostStats as mutable.  Making the stats() method non-const seems challenging due to widespread use of HostConstSharedPtr in load balancers.  Calls to stats() on a HostConstSharedPtr followed by increment calls on a counter are widespread, see source/common/router/router.cc for some examples.

Should we consider switch from std::shared_ptr<const Host> to std::shared_ptr<Host> as the primary way to reference Host objects in order to avoid use of mutable?

Reply:
> Should we consider switch from std::shared_ptr to std::shared_ptr as the primary way to reference Host objects in order to avoid use of mutable?

I'm not sure what the best thing to do here is. This is a pre-existing problem elsewhere, so it's not something you introduced, it's just never thrilled me that we ended up here and I wanted to make sure it's the same issue. I would say let's just merge this and we can consider further cleanups later if you are so inclined?

Reply:
Sounds good.  Thanks!